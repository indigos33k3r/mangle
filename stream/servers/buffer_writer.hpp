#ifndef MANGLE_STREAM_BUFFERWRITER_H
#define MANGLE_STREAM_BUFFERWRITER_H

#include "memory_stream.hpp"
#include <vector>

namespace Mangle {
namespace Stream {

/** A stream that writes to a contiguous memory buffer.

    The buffer is grown in size as needed, but it is highly
    recommended to pre-allocate the buffer using reserve() if you know
    the size up front.

    The class inherits from MemoryStream, so you can use it as a fully
    enabled read-write stream. Note the following caveats though:

    - Mangle does not separate write and read positions. So if you
      read and write to the same stream, make sure to keep track of
      the position yourself.

    - Writing to the stream, including using the getWPtr() functions,
      will invalidate all previous pointers returned from the
      stream. It also invalidates all streams obtained through
      MemoryStream::clone().
 */
class BufferWriter : public MemoryStream
{
  std::vector<char> buf;

 public:
  BufferWriter()
  {
    isWritable = true;
  }

  /** Get a WRITABLE pointer for 'count' bytes from the given
      position. Will allocate more memory as necessary. Does not
      affect the position pointer.
  */
  void *getWPtr(size_t pos, size_t count)
  {
    assert(buf.size() == size());

    size_t end = pos + count;
    if(end > buf.size())
      {
        // We needs us sum more buffer!
        buf.resize(end);

        // Reset data pointer and size
        size_t tmp = tell();
        set(&buf[0], buf.size());
        seek(tmp);

        assert(buf.size() == size());
        assert(size() == end);
      }

    // Return the pointer
    assert(end <= buf.size());
    return &buf[0] + pos;
  }

  /** Get a WRITABLE pointer for 'count' bytes from the current
      position. Like getPtr(count), this increases the position pointer
      as well.
  */
  void *getWPtr(size_t count)
  {
    void *res = getWPtr(tell(), count);
    seek(tell() + count);
    return res;
  }

  size_t write(const void* data, size_t len)
  {
    if(!len) return 0;

    // Copy the data
    memcpy(getWPtr(len), data, len);
    return len;
  }

  void reserve(size_t count)
  {
    buf.reserve(tell() + count);
  }
};

typedef boost::shared_ptr<BufferWriter> BufferWriterPtr;

}} // namespaces
#endif
