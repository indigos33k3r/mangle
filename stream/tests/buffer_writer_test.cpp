#include <iostream>
#include "../servers/buffer_writer.hpp"

using namespace Mangle::Stream;
using namespace std;

void info(Stream &str)
{
  cout << "Pos/Size: " << str.tell() << " / " << str.size();
  cout << "   Value: '" << string((const char*)str.getPtr(), str.size()) << "'\n\n";
}

void write(Stream &str, const std::string &msg)
{
  cout << "Writing '" << msg << "' (" << msg.size() << " bytes)\n";
  str.write(msg.c_str(), msg.size());
}

BufferWriter wrt;

void test(const std::string &msg)
{
  write(wrt, msg);
  info(wrt);
}

void testPos(const std::string &msg, int pos)
{
  cout << "Pointer-writing '" << msg << "' at position " << pos << endl;
  char *p = (char*)wrt.getWPtr(pos, msg.size());
  msg.copy(p, msg.size());
  info(wrt);
}

int main()
{
  assert(wrt.isReadable && wrt.isWritable && wrt.isSeekable &&
         wrt.hasPosition && wrt.hasSize && wrt.hasPtr);

  info(wrt);
  test("abc");
  test("def");
  test("1234567890");
  cout << "Seeking to 8\n";
  wrt.seek(8);
  test("blah");
  test("12345");
  testPos("ABCD", 1);
  testPos("9999", 17);
  test("xxx");

  return 0;
}
