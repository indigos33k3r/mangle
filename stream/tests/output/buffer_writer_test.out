Pos/Size: 0 / 0   Value: ''

Writing 'abc' (3 bytes)
Pos/Size: 3 / 3   Value: 'abc'

Writing 'def' (3 bytes)
Pos/Size: 6 / 6   Value: 'abcdef'

Writing '1234567890' (10 bytes)
Pos/Size: 16 / 16   Value: 'abcdef1234567890'

Seeking to 8
Writing 'blah' (4 bytes)
Pos/Size: 12 / 16   Value: 'abcdef12blah7890'

Writing '12345' (5 bytes)
Pos/Size: 17 / 17   Value: 'abcdef12blah12345'

Pointer-writing 'ABCD' at position 1
Pos/Size: 17 / 17   Value: 'aABCDf12blah12345'

Pointer-writing '9999' at position 17
Pos/Size: 17 / 21   Value: 'aABCDf12blah123459999'

Writing 'xxx' (3 bytes)
Pos/Size: 20 / 21   Value: 'aABCDf12blah12345xxx9'

